### PR TITLE
buildEnv: builder.pl: use signature and move the `ignoreSingleFileOutputs` parameter after the collision-related ones

### PR DESCRIPTION
**Pull request by ShamrockLee**
This PR uses Perl's `signatures` feature to increase readability and maintainability. This makes errors related to functional argument passing more discoverable.

Perl's `signatures` feature was added in Perl 5.20 and stabilized in Perl 5.32. As the Perl version in Nixpkgs `release-24.05` (the previous stable release branch) is already `5.38.2`, it should be safe to use such a feature in current Nixpkgs.

This PR also includes a commit to move the `ignoreSingleFileOutputs` argument after the ones that deal with collisions (to be exact, after `priority`). I added right after `ignoreCollisions` in #353752 for the convenience of my development but realized later that it would look awkward when reading through the function call.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by ShamrockLee** [comment]
I'll add some package tests for `buildEnv` to make the `builder.pl` changes safer. Before that, I'm marking this PR as a draft.

**Comment by philiptaron** [comment]
I added @stigtsp (who is busy and under no compulsion to review) as an FYI for Perl expertise.

**Review by philiptaron — COMMENTED** [review]
Adding tests is a good idea. The Perl passes my reading of it and a quick build on Big Computer. The parameter order changes to keep collision-related parameters together make sense.

**Review by stigtsp — APPROVED** [review]
Code LGTM

Using signatures is a great idea. This feature has in practice been stable for a long time and is commonly used.

**Review by philiptaron — APPROVED** [review]
Actually, while a test would be great, I would also be fine merging as-is.

File: pkgs/build-support/buildenv/builder.pl [hunk 1]
```diff
@@ -1,6 +1,7 @@
 #! @perl@ -w
 
 use strict;
```
[the remark this answers is not in the document]
Comment (ShamrockLee) [reply pkgs/build-support/buildenv/builder.pl:3]: Since Perl is at `5.40.0` on the stable branch, it should be safe to add such a requirement.

**Comment by ShamrockLee** [comment]
@stigtsp  Just tested `vim.customize { vimrcConfig = { }; }` and got some Perl errors:
```
vim> Bareword filehandle "DIR" not allowed under 'no feature "bareword_filehandles"' at /nix/store/dfgbszcdkp1i3ksbh4gpyify183v122p-builder.pl line 78.
vim> Bareword filehandle "DIR" not allowed under 'no feature "bareword_filehandles"' at /nix/store/dfgbszcdkp1i3ksbh4gpyify183v122p-builder.pl line 79.
vim> Bareword filehandle "DIR" not allowed under 'no feature "bareword_filehandles"' at /nix/store/dfgbszcdkp1i3ksbh4gpyify183v122p-builder.pl line 80.
vim> Bareword filehandle "PROP" not allowed under 'no feature "bareword_filehandles"' at /nix/store/dfgbszcdkp1i3ksbh4gpyify183v122p-builder.pl line 205.
vim> Bareword filehandle "PROP" not allowed under 'no feature "bareword_filehandles"' at /nix/store/dfgbszcdkp1i3ksbh4gpyify183v122p-builder.pl line 206.
vim> Bareword filehandle "PROP" not allowed under 'no feature "bareword_filehandles"' at /nix/store/dfgbszcdkp1i3ksbh4gpyify183v122p-builder.pl line 207.
vim> Bareword filehandle "FILE" not allowed under 'no feature "bareword_filehandles"' at /nix/store/dfgbszcdkp1i3ksbh4gpyify183v122p-builder.pl line 219.
vim> Bareword filehandle "FILE" not allowed under 'no feature "bareword_filehandles"' at /nix/store/dfgbszcdkp1i3ksbh4gpyify183v122p-builder.pl line 220.
vim> Bareword filehandle "FILE" not allowed under 'no feature "bareword_filehandles"' at /nix/store/dfgbszcdkp1i3ksbh4gpyify183v122p-builder.pl line 221.
vim> Bareword filehandle "FILE" not allowed under 'no feature "bareword_filehandles"' at /nix/store/dfgbszcdkp1i3ksbh4gpyify183v122p-builder.pl line 255.
vim> /nix/store/dfgbszcdkp1i3ksbh4gpyify183v122p-builder.pl has too many errors.
```

**Comment by ShamrockLee** [comment]
@stigtsp Since `use v5.40` requires extra fixes, let's do it in a separate PR instead. I'll drop the commit from this PR.

**Comment by ShamrockLee** [comment]
The change has been reset, and `vim.customize { vimrcConfig = { }; }` now builds.

@philiptaron, should we merge now?

File: pkgs/build-support/buildenv/builder.pl (hunk 1 above)
[the remark this answers is not in the document]
Comment (ShamrockLee) [reply pkgs/build-support/buildenv/builder.pl:3]: Let's do it in a new PR instead.

**Comment by stigtsp** [comment]
> @stigtsp Since `use v5.40` requires extra fixes, let's do it in a separate PR instead. I'll drop the commit from this PR.

If it doesn't work with the feature flags from 5.40, then it's fine to not include that change of course. :+1:

**Pull request merged**